In [2]:
!pip3 install pandas dash

In [3]:
import pandas as pd
import dash
import dash_html_components as html
import dash_core_components as dcc
from dash.dependencies import Input, Output, State
import plotly.graph_objects as go
import plotly.express as px
from dash import no_update

<ipython-input-3-02b83762fc9e>:3: UserWarning: 
The dash_html_components package is deprecated. Please replace
`import dash_html_components as html` with `from dash import html`
  import dash_html_components as html
<ipython-input-3-02b83762fc9e>:4: UserWarning: 
The dash_core_components package is deprecated. Please replace
`import dash_core_components as dcc` with `from dash import dcc`
  import dash_core_components as dcc


In [4]:
import wget

In [134]:
# Download and read the `spacex_launch_geo.csv`
spacex_csv_file = wget.download('https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/datasets/spacex_launch_geo.csv')
spacex_df=pd.read_csv(spacex_csv_file)

In [23]:
spacex_df.head()

Flight Number        Date Time (UTC) Booster Version  Launch Site  \
0              1  2010-06-04   18:45:00  F9 v1.0  B0003  CCAFS LC-40   
1              2  2010-12-08   15:43:00  F9 v1.0  B0004  CCAFS LC-40   
2              3  2012-05-22    7:44:00  F9 v1.0  B0005  CCAFS LC-40   
3              4  2012-10-08    0:35:00  F9 v1.0  B0006  CCAFS LC-40   
4              5  2013-03-01   15:10:00  F9 v1.0  B0007  CCAFS LC-40   

                                             Payload  Payload Mass (kg)  \
0               Dragon Spacecraft Qualification Unit                0.0   
1  Dragon demo flight C1, two CubeSats,  barrel o...                0.0   
2                             Dragon demo flight C2+              525.0   
3                                       SpaceX CRS-1              500.0   
4                                       SpaceX CRS-2              677.0   

       Orbit         Customer        Landing Outcome  class        Lat  \
0        LEO           SpaceX  Failure   (parachute)      0  28.562302   
1  LEO (ISS)  NASA (COTS) NRO  Failure   (parachute)      0  28.562302   
2  LEO (ISS)      NASA (COTS)             No attempt      0  28.562302   
3  LEO (ISS)       NASA (CRS)             No attempt      0  28.562302   
4  LEO (ISS)       NASA (CRS)             No attempt      0  28.562302   

        Long  
0 -80.577356  
1 -80.577356  
2 -80.577356  
3 -80.577356  
4 -80.577356

In [24]:
spacex_df.columns

Index(['Flight Number', 'Date', 'Time (UTC)', 'Booster Version', 'Launch Site',
       'Payload', 'Payload Mass (kg)', 'Orbit', 'Customer', 'Landing Outcome',
       'class', 'Lat', 'Long'],
      dtype='object')

In [8]:
fig1_df = spacex_df[['Launch Site', 'class']]
fig1_df = fig1_df.groupby(['Launch Site'], as_index=False).count()
fig1_df

Launch Site  class
0   CCAFS LC-40     26
1  CCAFS SLC-40      7
2    KSC LC-39A     13
3   VAFB SLC-4E     10

In [83]:
filtro = spacex_df[spacex_df['Launch Site']=='KSC LC-39A'].groupby(['Launch Site'], as_index=False).count()
filtro

Launch Site  Flight Number  Date  Time (UTC)  Booster Version  Payload  \
0  KSC LC-39A             13    13          13               13       13   

   Payload Mass (kg)  Orbit  Customer  Landing Outcome  class  Lat  Long  
0                 13     13        13               13     13   13    13

In [114]:
fig1 = px.pie(spacex_df, values='class', names='Launch Site', title="Total Success Launches By Site")
fig1.show()

In [124]:
fig3 = px.scatter(spacex_df, x='Payload Mass (kg)', y='class', title='Correlation between payload and Success for all sites', 
                  color = 'Booster Version')
fig3.show()

In [116]:
max_payload = spacex_df['Payload Mass (kg)'].max()
min_payload = spacex_df['Payload Mass (kg)'].min()
print(max_payload)
print(min_payload)

9600.0
0.0


In [117]:
# Create Layout 
app.layout = html.Div( children = [
    html.H1('SpaceX Launch Records Dashboard', style={'textAlign': 'center', 'color': '#503D36','font-size': 40}),

    # Create dropdown
    dcc.Dropdown(options=[
            {'label': 'All Sites', 'value': 'ALL'},
            {'label': 'CCAFS LC-40', 'value': 'CCAFS LC-40'},
            {'label': 'CCAFS SLC-40', 'value': 'CCAFS SLC-40'},
            {'label': 'KSC LC-39A', 'value': 'KSC LC-39A'},
            {'label': 'VAFB SLC-4E', 'value': 'VAFB SLC-4E'}  
            ],        
            value='ALL' # Providing a vallue to dropdown
    ),
    
    # Bar graph
    dcc.Graph(id='fig-pie',figure=fig1),
    
    dcc.RangeSlider(id='range-slider', min=0,max=10000, 
                                marks={0: '0 kg', 2500: '2500 kg', 5000: '5000 kg',7500: '7500 kg', 10000: '10000 kg' },
                                value=[min_payload, max_payload],allowCross=False),

    
    dcc.Graph(id='fig-scatter',figure=fig3),


])

In [119]:
df_launch = spacex_df[['Launch Site','class']].groupby(['Launch Site']).mean('class')
df_launch

class
Launch Site           
CCAFS LC-40   0.269231
CCAFS SLC-40  0.428571
KSC LC-39A    0.769231
VAFB SLC-4E   0.400000

In [ ]:
app = dash.Dash(__name__)

# REVIEW1: Clear the layout and do not display exception till callback gets executed
app.config.suppress_callback_exceptions = True


app.layout = html.Div(children=[html.H1('SpaceX Launch Records Dashboard',
                                        style={'textAlign': 'center', 'color': '#503D36',
                                               'font-size': 40}),

                                dcc.Dropdown(id='site-dropdown',
                                             options=[
                                                     {'label': 'All Sites', 'value': 'ALL'},
                                                     {'label': 'CCAFS LC-40', 'value': 'CCAFS LC-40'},
                                                     {'label': 'VAFB SLC-4E', 'value': 'VAFB SLC-4E'},
                                                     {'label': 'KSC LC-39A', 'value': 'KSC LC-39A'},
                                                     {'label': 'CCAFS SLC-40', 'value': 'CCAFS SLC-40'}
                                                     ],
                                             value='ALL',
                                             placeholder='Select a Launch Site here',
                                             searchable=True
                                            ),
                                html.Br(),

                                html.Div(dcc.Graph(id='fig-pie')),
                                html.Br(),

                                html.P("Payload range (Kg):"),
                                html.Br(),
                                
                                dcc.RangeSlider(id='payload-slider',
                                                min=0,
                                                max=10000,
                                                step=2500,
                                                value=[min_payload, max_payload]
                                                ),

                                html.Div(dcc.Graph(id='fig-scatter')),
                                ])

@app.callback(Output(component_id='fig-pie', component_property='figure'),
              Input(component_id='site-dropdown', component_property='value'))
def get_pie_chart(entered_site):
    filtered_df = spacex_df
    if entered_site == 'ALL':
        fig = px.pie(filtered_df, values='class', 
        names='Launch Site', 
        title='Success Count for all launch sites')
        return fig
    else:
        # return the outcomes piechart for a selected site
        filtered_df=spacex_df[spacex_df['Launch Site']== entered_site]
        filtered_df=filtered_df.groupby(['Launch Site','class']).size().reset_index(name='class count')
        fig=px.pie(filtered_df,values='class count',names='class',title=f"Total Success Launches for site {entered_site}")
        return fig

# TASK 4:
# Add a callback function for `site-dropdown` and `payload-slider` as inputs, `success-payload-scatter-chart` as output
@app.callback(Output(component_id='fig-scatter',component_property='figure'),
                [Input(component_id='site-dropdown',component_property='value'),
                Input(component_id='payload-slider',component_property='value')])
def scatter(entered_site,payload):
    filtered_df = spacex_df[spacex_df['Payload Mass (kg)'].between(payload[0],payload[1])]
    # thought reusing filtered_df may cause issues, but tried it out of curiosity and it seems to be working fine
    
    if entered_site=='ALL':
        fig=px.scatter(filtered_df,x='Payload Mass (kg)',y='class',color='Booster Version',title='Success count on Payload mass for all sites')
        return fig
    else:
        fig=px.scatter(filtered_df[filtered_df['Launch Site']==entered_site],x='Payload Mass (kg)',y='class',color='Booster Version',title=f"Success count on Payload mass for site {entered_site}")
        return fig
    
if __name__ == '__main__':
    app.run_server(port=8051)    

Dash is running on http://127.0.0.1:8051/

Dash is running on http://127.0.0.1:8051/

Dash is running on http://127.0.0.1:8051/

 * Serving Flask app "__main__" (lazy loading)
 * Environment: production
   Use a production WSGI server instead.
 * Debug mode: off


 * Running on http://127.0.0.1:8051/ (Press CTRL+C to quit)
127.0.0.1 - - [25/Jul/2022 05:08:02] "GET / HTTP/1.1" 200 -
127.0.0.1 - - [25/Jul/2022 05:08:02] "GET /_dash-layout HTTP/1.1" 200 -
127.0.0.1 - - [25/Jul/2022 05:08:02] "GET /_dash-dependencies HTTP/1.1" 200 -
127.0.0.1 - - [25/Jul/2022 05:08:02] "GET /_dash-component-suites/dash/dcc/async-dropdown.js HTTP/1.1" 200 -
127.0.0.1 - - [25/Jul/2022 05:08:02] "GET /_dash-component-suites/dash/dcc/async-graph.js HTTP/1.1" 200 -
127.0.0.1 - - [25/Jul/2022 05:08:02] "GET /_dash-component-suites/dash/dcc/async-slider.js HTTP/1.1" 200 -
127.0.0.1 - - [25/Jul/2022 05:08:02] "GET /_dash-component-suites/dash/dcc/async-plotlyjs.js HTTP/1.1" 200 -
127.0.0.1 - - [25/Jul/2022 05:08:03] "POST /_dash-update-component HTTP/1.1" 200 -
127.0.0.1 - - [25/Jul/2022 05:08:03] "POST /_dash-update-component HTTP/1.1" 200 -
127.0.0.1 - - [25/Jul/2022 05:08:10] "POST /_dash-update-component HTTP/1.1" 200 -
127.0.0.1 - - [25/Jul/2022 05:08:14] "POST /_dash-

    Which site has the largest successful launches? KSC LC-39A, com 10 sucessos
    
    Which site has the highest launch success rate? KSC LC-39A, com 77% de taxa de sucesso
    
    Which payload range(s) has the highest launch success rate? Payload entre 2500 e 5000kg é o range onde se encontra maior sucesso, com 11 lançamentos com sucesso em relação aos sucessos dos demais ranges do Payload.
    
    Which payload range(s) has the lowest launch success rate? O range de 7500 à 10000Kg com apenas 1 sucesso.
    
    Which F9 Booster version (v1.0, v1.1, FT, B4, B5, etc.) has the highest launch success rate???